drv/grapefruit-seq-server/src/main.rs

```diff
@@ -0,0 +1,266 @@
+            let mut buffer = [0; idl::INCOMING_SIZE];
+            loop {
+                idol_runtime::dispatch(&mut buffer, &mut server);
+            }
+        }
+
+        // Initializing the sequencer failed.
+        Err(e) => {
+            // Tell everyone that something's broken, as loudly as possible.
+            ringbuf_entry!(Trace::StartFailed(e));
+            sys.gpio_set(FAULT_PIN_L); // TODO should this be reset, not set?
+
+            // All these moments will be lost in time, like tears in rain...
+            // Time to die.
+            loop {
+                // Sleeping with all bits in the notification mask clear means
+                // we should never be notified --- and if one never wakes up,
+                // the difference between sleeping and dying seems kind of
+                // irrelevant. But, `rustc` doesn't realize that this should
+                // never return, we'll stick it in a `loop` anyway so the main
+                // function can return `!`
+                sys_recv_notification(0);
+            }
+        }
+    }
+}
+
+struct ServerImpl<S: SpiServer> {
+    sys: sys_api::Sys,
+    seq: SpiDevice<S>,
+}
+
+const FAULT_PIN_L: sys_api::PinSet = sys_api::Port::A.pin(15);
+
+const FPGA_PROGRAM_L: sys_api::PinSet = sys_api::Port::B.pin(5);
+const FPGA_INIT_L: sys_api::PinSet = sys_api::Port::B.pin(6);
+const FPGA_CONFIG_DONE: sys_api::PinSet = sys_api::Port::B.pin(4);
+
+impl<S: SpiServer + Clone> ServerImpl<S> {
+    fn init(
+        sys: &sys_api::Sys,
+        spi: S,
+        aux: drv_auxflash_api::AuxFlash,
+    ) -> Result<Self, SeqError> {
+        // Ensure the SP fault pin is configured as an open-drain output, and pull
+        // it low to make the sequencer restart externally visible.
+        sys.gpio_configure_output(
+            FAULT_PIN_L,
+            sys_api::OutputType::OpenDrain,
+            sys_api::Speed::Low,
+            sys_api::Pull::None,
+        );
+        sys.gpio_reset(FAULT_PIN_L);
+
+        // To allow for the possibility that we are restarting, rather than
+        // starting, we take care during early sequencing to _not turn anything
+        // off,_ only on. This means if it was _already_ on, the outputs should
+        // not glitch.
+
+        // To program the FPGA, we're using "slave serial" mode.
+        //
+        // See "7 Series FPGAs Configuration", UG470 (v1.17) for details,
+        // as well as "Using a Microprocessor to Configure Xilinx 7 Series FPGAs
+        // via Slave Serial or Slave SelectMAP Mode Application Note" (XAPP583)
+
+        // Configure the PROGRAM_B line to the FPGA
+        sys.gpio_set(FPGA_PROGRAM_L);
+        sys.gpio_configure_output(
+            FPGA_PROGRAM_L,
+            sys_api::OutputType::OpenDrain,
+            sys_api::Speed::Low,
+            sys_api::Pull::None,
+        );
+
+        // Pulse PROGRAM_B low for 1 ms to reset the bitstream
+        // (T_PROGRAM is 250 ns, so this is fine)
```

Comment:
i'm assuming that this means that 250 ns is the _minimum_?

Reply:
Yup, updated the comment